src/anchor-navigation/internal.tsx

```diff
@@ -0,0 +1,117 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import React, { useCallback, useEffect, useMemo } from 'react';
+import clsx from 'clsx';
+import styles from './styles.css.js';
+import { AnchorNavigationProps } from './interfaces';
+import { checkSafeUrl } from '../internal/utils/check-safe-url';
+import useScrollSpy from './scroll-spy.js';
+import { fireCancelableEvent, fireNonCancelableEvent, isPlainLeftClick } from '../internal/events/index';
+import { useStableCallback } from '@cloudscape-design/component-toolkit/internal';
+import { InternalBaseComponentProps } from '../internal/hooks/use-base-component/index.js';
+import { getBaseProps } from '../internal/base-component/index.js';
+
+export default function InternalAnchorNavigation({
+  anchors,
+  ariaLabelledby,
+  onFollow,
+  onActiveAnchorChange,
+  disableTracking = false,
+  activeHref = '',
+  __internalRootRef = null,
+  ...props
+}: AnchorNavigationProps & InternalBaseComponentProps) {
+  const baseProps = getBaseProps(props);
+
+  const hrefs = useMemo(() => anchors.map(anchor => anchor.href), [anchors]);
+  const memoizedAnchors = useMemo(() => anchors, [anchors]);
+
+  const onFollowHandler = useCallback(
+    (anchor: AnchorNavigationProps.Anchor, sourceEvent: React.SyntheticEvent | Event) => {
+      fireCancelableEvent(onFollow, anchor, sourceEvent);
+    },
+    [onFollow]
+  );
+
+  const onActiveAnchorChangeHandler = useStableCallback((newActiveAnchor: AnchorNavigationProps.Anchor | undefined) => {
+    fireNonCancelableEvent(onActiveAnchorChange, newActiveAnchor);
+  });
+
+  const [currentActiveHref, setCurrentActiveHref, setDisableTracking] = useScrollSpy({ hrefs });
+
+  useEffect(() => {
+    setDisableTracking(disableTracking);
+
+    if (activeHref) {
+      setCurrentActiveHref(activeHref);
+    }
+  }, [setDisableTracking, setCurrentActiveHref, disableTracking, activeHref]);
+
+  useEffect(() => {
+    if (currentActiveHref) {
+      onActiveAnchorChangeHandler(memoizedAnchors.find(anchor => anchor.href === currentActiveHref));
```

Comment:
It would be nice to have a test that asserts that `onActiveAnchorChange` is called once and only once on anchor change.

Reply:
TODO